readme.md
## Massive-Collections

Collections wrapper for Massive-JS.
You must already have a connection and Massive-Collections work with your database connection object.

## Dependencies

* bluebird
* massive (connection)

Let's admit we have a users table :

```sql
(
  id       serial primary key
  name     varchar(255),
  age      smallint,
  infos    jsonb not null,
  created  timestamp with time zone default now(),
  modified timestamp with time zone default now()
)
```

Create a collection file : `users.js`


```javascript
const { ColumnMissing } = require('massive-collections/errors');
const Collection = require('massive-collections');

// Get the connection from massive
module.exports = (db) => {

  const UsersCollection = new Collection('users', db);

  return UsersCollection;
};
```

## Methods

Now you can use following methods :

* get (id)
* count
* find
* insert
* update
* updateAll
* remove
* removeAll
* flush

Each method returns a Promise.

### Count method

Purpose: Count database row.

Returns: {Number}

| Parameter  | Type   | Description                               | Example                                                                                   |
|:-----------|:-------|:------------------------------------------|:------------------------------------------------------------------------------------------|
| conditions | Object | WHERE conditions                          | { "name ~~": "jo%"  }  // name like                                                       |

`Collection.count(conditions)`

```javascript

  Users.count().then(res => {
    console.log(res);
  });

```

### Get method

Purpose: Get a specific row.

Returns: {Object}

| Parameter  | Type   | Description      | Example  |
|:-----------|:-------|:-----------------|:---------|
| id         | Number | ID               | 5        |

`Collection.get(id)`

```javascript

  Users.get(10).then(user => {
    console.log(user);
  });

```

### Find method

Purpose: Get an array of rows.

Returns: {Array}

| Parameter  | Type   | Description                               | Example                                                                                   |
|:-----------|:-------|:------------------------------------------|:------------------------------------------------------------------------------------------|
| conditions | Object | WHERE conditions                          | { "name ~~": "jo%"  }  // name like                                                       |
| options    | Object | Other options, like limit, offset, etc... | { columns: ["name", "price", "description"], order: {field: "price", direction: "desc"}, offset: 20, limit: 10 } |

`Collection.find(conditions, options)`

```javascript

  Users.find().then(res => {
    console.log(res);
  });

```

### Insert method

Purpose: Insert an new row in our table.

Returns: {Object}

| Parameter  | Type   | Description      | Example                                                        |
|:-----------|:-------|:-----------------|:---------------------------------------------------------------|
| data       | Object | Values           | { name: "John Doe", infos: { email: "john.doe@domain.tld" } }  |

`Collection.insert(data)`

```javascript

  UsersCollection.insert({
    name: "Jane Doe",
    infos: {
      email: "jane.doe@domain.tld"
    }
  }).then(res => {
    console.log(res);
  });

```

### Update method

Purpose: Update a row where id = ...

Returns: {Object} (updated row)

| Parameter  | Type   | Description      | Example                                                        |
|:-----------|:-------|:-----------------|:---------------------------------------------------------------|
| id         | Number | id of the item   | 3                                                              |
| data       | Object | new data to set  | { name: "bobby" }                                              |

`Collection.update(id, data)`

```javascript

  UsersCollection.update(11, {
    name: "Toto"
  }).then(res => {
    console.log(res);
  });

```

### UpdateAll method

Purpose: Update any rows where conditions match

Returns: {Array} (updated rows)

| Parameter  | Type   | Description      | Example                                                        |
|:-----------|:-------|:-----------------|:---------------------------------------------------------------|
| conditions | Object | WHERE conditions | { "name ~~": "jo%"  }  // name like                            |
| data       | Object | new data to set  | { name: "bobby" }                                              |

`Collection.updateAll(conditions, data)`

```javascript

  UsersCollection.updateAll({
    'name ilike': 't%' // Find all name starting with t non case sensitive
  }, {
    age: 20 // Set age = 20
  }).then(res => {
    console.log(res);
  });

```

### Remove method

Purpose: Remove a row where id = ...

Returns: {Object} (deleted item)

| Parameter  | Type   | Description      | Example     |
|:-----------|:-------|:-----------------|:------------|
| id         | Number | id of the item   | 5           |

`Collection.remove(id)`

```javascript

  UsersCollection.remove(5).then(res => {
    console.log(res);
  });

```

### RemoveAll method

Purpose: Remove any rows that match conditions

Returns: {Array} (deleted items)

| Parameter  | Type   | Description      | Example                             |
|:-----------|:-------|:-----------------|:------------------------------------|
| conditions | Object | WHERE conditions | { "name ~~": "jo%"  }  // name like |

`Collection.removeAll(conditions)`

```javascript

  UsersCollection.removeAll({
    'username ilike': 'jo%'
  }).then(res => {
    console.log(res);
  });

```

### Flush method

Purpose: Remove any rows in that table

| Parameter  | Type    | Description                                                         | Example    |
|:-----------|:--------|:--------------------------------------------------------------------|:-----------|
| reset_seq  | Boolean | (Optionnal) Describe if we need to reset the linked sequence or not | true|false |

`Collection.flush(reset_seq)`

```javascript

  UsersCollection.flush().then(()  => {

  });

  // Reset sequence
  UsersCollection.flush(true).then(() => {

  })

```

## Formatters

You can format data when you read/write.

**dbFormat** before any write
**jsFormat** before any read

This will be helpful with custom format or extra libs (password hash, etc...).

```javascript

UsersCollection.dbFormat(data => {

  if (typeof data.infos === "undefined")
    throw new ColumnMissing(infos);

  // Force data convertion
  if (typeof data.infos === "object")
    data.infos = JSON.stringify(data.infos);

  return data;
});

```

## Hooks

You can hook methods call and change data value before any database write.

To assign a pre hook, we use the following method : `Collection.preHook(hookName, callback)`

```javascript
UsersCollection.preHook('update', function (next, data) {
  data.modified = new Date();
  next(data);
});

```

To assign a post hook, we use the following method : `Collection.postHook(hookName, callback)`

```javascript
UsersCollection.postHook('update', function (next, data) {
  data.new_field = 'qwerty';
  myRestartFunction(); // any action
  next(data);
});
```

### List of Hooks

* pre->count(next)
* pre->get(next)
* pre->find(next)
* pre->flush(next)
* pre->insert(next, data)
* pre->update(next, data)
* pre->updateAll(next, data)
* pre->remove(next)
* post->get(data)
* post->count(data)
* post->find(data)
* post->flush()
* post->insert(data)
* post->update(data)
* post->updateAll(data)
* post->remove(data)

For Pre Insert and Pre Update, you must pass `data` through the `next` callback.

### Custom queries

**Like** : **~~** :

```javascript
UsersCollection.find({
  "name ~~": 'ja%'
}).then(res => console.log(res));
```

**Not Like** : **!~~** :

```javascript
UsersCollection.find({
  "name !~~": 'ja%'
}).then(res => console.log(res));
```

**iLike** (case insensitive) :

```javascript
UsersCollection.find({
  "name ilike": '%joh%'
}).then(res => console.log(res));
```

**Not iLike** (case insensitive) :

```javascript
UsersCollection.find({
  "name not ilike": '%joh%'
}).then(res => console.log(res));
```

**Compare** : **>** **<** **<=** **>=** :
```javascript
UsersCollection.find({
  "age >": 30
}).then(res => console.log(res));
```

**JSONB queries** :

Let's assume that we have a columns `infos` :
```
{ "email": "...", "followers": 5000 }
```

```javascript
// Value in object like
UsersCollection.find({
  "infos->>'email' ~~": "jo%"
}).then(res => console.log(res));

// Get value then cast
UsersCollection.find({
  "(infos->>'followers')::float >": 600
}).then(res => console.log(res));

// Sort
UsersCollection.find({}, {
  order: [{
    field: "infos->>'followers'",
    direction: "DESC",
    type: "int"
  }]
}).then(res => console.log(res));
```
## CLI

You can create table from a terminal with massive-collections-cli. Let's assume that you already have a connection to a postgresql database (user, password, etc.).

You need to connect first (in a terminal) :

```
node_modules/.bin/massive-collections-cli connect --h=localhost:5432 --db=test_db --u=root --p=root
```

Once you are connected, you generate automatically a new file : `massive-collections_credentials.json` that should automatically be added to your .gitignore.

Then, you can create tables (in a terminal) :

```
# Note that we use double quote to prevent bash errors
node_modules/.bin/massive-collections-cli createTable posts "title:varchar(255):unique:notnull" "content:text" "picture:integer" "author:integer" "details:jsonb" "created:timestampz:noindex:null:now()"
```


Do not add an id column, this is automatic.

If you want to remove properly your credentials, you can disconnect (in a terminal) :

```
node_modules/.bin/massive-collections-cli disconnect
```

Please read the documentation first (in a terminal) :

```
node_modules/.bin/massive-collections-cli help
```


changelog.txt
# v1.3.0

- Fix config path exclusion

================================================================================

# v1.2.9

- Fix Promise returns

================================================================================

# v1.2.7

- Move dbFormat before real database query to be the last action (insert, update, updateAll)

================================================================================

# v1.2.6

- Changing post hooks to promises

================================================================================

# v1.2.5

- Reset sequence on flush

================================================================================

# v1.2.4

- Fixing date for updateAll

================================================================================

# v1.2.1

- Fixing escape char

================================================================================

# v1.2.0

- Starting hooks tests (not finished)

================================================================================

# v1.1.9

- Collection.removeAll method
- Hook pre/post removeAll

================================================================================

# v1.1.8

- Collection.updateAll method
- Hook pre/post updateAll

================================================================================

# v1.1.7

- Collection.count method

================================================================================

# v1.1.6

- Working CLI
- Collection.get method

================================================================================

# Older

- Collections features :

  * find
  * insert
  * update
  * remove

- Hooks :

  * pre
  * post

- Formatters :

  * toJS
  * toDB


tests/index.js
// Test Massive-Collections
require('./massive-collections.test');

// Test CLI
require('./massive-collections-cli.test');

tests/config.js
module.exports = {
  host: 'localhost',
  port: 5432,
  database: 'testdb',
  user: 'quizeruser',
  password: 'abobo10'
};


massive-collections.test.js
const massive = require('massive');
const Collection = require('./index');
const config = require('./config');
const fs = require('fs');


describe('Massive-Collections tests', () => {

  // Init our db connection
  beforeAll((done) => {
    massive(config).then(db => {
      process.db = db;
      done();
    })
  });

  test('Should create a fake table in our database', (done) => {
    const fakeTableQuery = fs.readFileSync('./faketable.sql', 'utf8');

    process.db.query(fakeTableQuery).then(() => {
      const FakeTable = new Collection('fake_table', process.db);

      FakeTable.flush().then(() => {

        FakeTable.find({}, {}).then(res => {
          expect(res.length).toBe(0);
          done();
        });

      });
    });
  });

  test('Should insert data', (done) => {
    const FakeTable = new Collection('fake_table', process.db);

    FakeTable.insert({
      username: 'John Doe',
      password: 'qwerty'
    }).then(() => {

      FakeTable.find({}).then(res => {
        expect(res.length).toBe(1);
        expect(res[0].username).toBe('John Doe');
        expect(res[0].password).toBe('qwerty');
        done();
      });
    });
  });


  test('Should clean data', (done) => {
    const FakeTable = new Collection('fake_table', process.db);

    FakeTable.flush().then(() => {
      FakeTable.find({}).then(res => {
        expect(res.length).toBe(0);
        done();
      });
    });
  });
})


help.js
const log = require('color-log');

log.info("\n====================================\n");
log.info("\n| How To Use Massive-Collection CLI |\n");
log.info("\n====================================\n");

log.mark("\nAvailable commands : \n");

log.info(`
  * connect                     Connect to Database (permanent)
  * disconnect                  Disconnect
  * create-table | createTable  Create a table
          `);

log.mark("\nParameters : \n");

log.info(`
  * connect [parameters]

  list :
  --h     host:port
  --db    database
  --p     password
  --u     user
  `);

log.warn('Example : massive-collections-cli connect --h=localhost:5432 --db=test_db --u=root --p=root');

log.info(`
  * createTable <tableName> <...columns>

  column format:
  name:type:[index]:[nullable]:[default]

  * Details :
  name      Any
  type      Any postgresql type (or shortcuts : timestampz, int, bool)
  index     <unique|noindex> (optionnal)
  nullable  <null|notnull> (optionnal)
  default   <now()|true|false|...> (optionnal)
  `);

log.warn(`
  Example :
  massive-collections createTable users \\
  username:varchar(255):unique:notnull \\
  password:varchar(255):noindex:notnull \\
  age:int \\
  details:jsonb \\
  created:timestampz:noindex:null:now() \\
  modified:timestampz:noindex:null:now()
`);

massive-collections-cli.js
const log = require('color-log');
const argv = require('minimist')(process.argv.slice(2));
const path = require('path');
const os = require('os');
const fs = require('fs');
const massive = require('massive');

const currentPath = path.resolve(__dirname);

// File that will contain our credentials
const credentialsPath = path.join(__dirname, 'massive-collections_credentials.json');

// Check git ignore
if(currentPath.indexOf('node_modules') > -1) {
  const parentPath = path.resolve(currentPath.split('node_modules')[0]);

  const gitIgnorePath = path.join(parentPath, '.gitignore');

  // Get relative path from gitignore file
  const credentialsRelative = credentialsPath.substring(parentPath.length, credentialsPath.length);

  if(fs.existsSync(gitIgnorePath) && fs.lstatSync(gitIgnorePath)) {
    const gitIgnoreContent = fs.readFileSync(gitIgnorePath, 'utf8');

    let splitted = gitIgnoreContent.split(os.EOL);

    let found = false;

    for(let i in splitted) {
      if(splitted[i].match(credentialsRelative + "\s*$")) {
        found = true;
      }
    }

    // If credientials is not in gitignore
    if(!found) {
      // Add it
      splitted.push(credentialsRelative);

      // Save
      fs.writeFileSync(gitIgnorePath, splitted.join(os.EOL));
    }
  }
}

/**
* @entry connect
* @type CLI
*
* Creates credientials file to execute queries
*
* @param {String} address host:port
* @param {String} db
* @param {String} user
* @param {String} password
* @constraint address must be type of string
* @constraint address be formatted like 'host:port'
* @constraint db must be type of string
* @constraint user must be type of string
* @constraint password must be type of string
*/
function connect(address, db, user, password) {
  if(typeof address === "undefined" || typeof user === "undefined" || typeof password === "undefined" || typeof db === "undefined")
    throw new Error('Missing parameter, address, db, user and password are required to connect.');

  const [host, port] = address.split(':');

  if(typeof host === "undefined" || typeof port === "undefined")
    throw new Error('Bad address format, it should be like host:port');

  const cred = { user, password, host, port, database: db };

  fs.writeFileSync(credentialsPath, JSON.stringify(cred));

  log.info('Connected');
}

/**
* @entry disconnect
* @type CLI
*
* Remove credientials file
*/
function disconnect() {
  if(fs.existsSync(credentialsPath) && fs.lstatSync(credentialsPath).isFile()) {
    fs.unlinkSync(credentialsPath);
  }

  log.info('Disconnected');
}

/**
* @entry createTable
* @type CLI
*
* Generate a create table query
*
* @param {String} tableName
* @param {Array} columns (column:type:[index]:[nullable]:[default])
* @constraint tableName must be type of string
* @constraint columns must be type of array
*/
function createTable(tableName, columns) {
  if(typeof tableName === "undefined" || typeof columns === "undefined")
    throw new Error('Missing parameter, table name and columns must be defined.');

  if(typeof columns.splice === "undefined")
    throw new Error('Columns must be type of array.');

  if(columns.length === 0)
    throw new Error('Columns missing.');

  const cols = ['id serial primary key'];

  // Sanitize types
  function handleType(type) {
    switch(type.toLowerCase()) {
      case "int":
        type = "integer";
      break;

      case "bool":
        type = "boolean";
      break;

      case "timestampz":
        type = "timestamp with time zone"
      break;

      default:
        return type;
    }

    return type;
  }

  // Sanitize index
  function handleIndex(index) {
    switch(index.toLowerCase()) {
      case "unique" :
        index = " UNIQUE";
      break;

      case "noindex" :
        index = "";
      break;

      default:
        throw new Error('Bad index value : ' + index);
      break;
    }

    return index;
  }

  // Sanitize nullable
  function handleNullable(nullable) {
    switch(nullable.toLowerCase()) {
      case "notnull":
        nullable = " NOT NULL";
      break;

      case "null":
        nullable = "";
      break;

      default:
        throw new Error('Bad nullable value : ' + nullable);
    }

    return nullable;
  }

  for(let i in columns) {
    let column = columns[i].split(':');

    if(column.length < 2)
      throw new Error('Bad column format : ' + columns[i]);

    switch(column.length) {
      case 2 :
        var [name, type] = column;

        type = handleType(type);

        cols.push(`${name} ${type}`);
      break;

      case 3 :
        var [name, type, index] = column;

        type = handleType(type);
        index = handleIndex(index);

        cols.push(`${name} ${type}${index}`);
      break;

      case 4 :
        var [name, type, index, nullable] = column;

        type = handleType(type);
        index = handleIndex(index);
        nullable = handleNullable(nullable);

        cols.push(`${name} ${type}${index}${nullable}`);
      break;

      case 5 :
        var [name, type, index, nullable, def] = column;

        if(def.length === 0)
          throw new Error('Missing default value for column ' + name);

        type = handleType(type);
        index = handleIndex(index);
        nullable = handleNullable(nullable);

        cols.push(`${name} ${type}${index}${nullable} DEFAULT ${def}`);
      break;
    }
  }

  let query = `CREATE TABLE IF NOT EXISTS "${tableName}" ( ${cols.join(', ')} )`;

  return query;
}

/**
* @entry runQuery
* @type CLI
*
* Connect to a db then execute a query
*
* @param {String} query
* @constraint query must be type of string
*/
function runQuery(query) {
  if(typeof query !== "string")
    throw new Error('Bad query');

  if(!fs.existsSync(credentialsPath))
    throw new Error('Error, you must connect first.');

  const credentials = require(credentialsPath);

  massive(credentials).then(db => {
    db.query(query).then((res) => {
      log.info('Done.');
      process.exit(0);
    }).catch(err => {
      throw err;
    })
  });
}

try {
  // Handle argv
  if(argv._.length > 0) {
    console.log(argv);

    switch(argv._[0]) {
      case 'connect':
        connect(argv.h, argv.db, argv.u, argv.p);
      break;

      case 'disconnect':
        disconnect();
      break;

      case 'create-table':
      case 'createTable':
        if(argv._.length < 3) 
          throw new Error('Error, not enough arguments.');

        const tableName = argv._[1];
        const cols = argv._.splice(2, argv._.length);

        const query = createTable(tableName, cols);
        
        runQuery(query);
      break;

      case 'help':
        log.info("\n====================================\n");
        log.info("\n| How To Use Massive-Collection CLI |\n");
        log.info("\n====================================\n");
        log.warn("\nAvailable commands : \n");
        log.info(`
  * connect                     Connect to Database (permanent)
  * disconnect                  Disconnect
  * create-table | createTable  Create a table
          `);
        log.warn("\nParameters : \n");
        log.info(`
  * connect [parameters]

  list :
  --h     host:port
  --db    database
  --p     password
  --u     user

  Example : massive-collections-cli connect --h=localhost:5432 --db=test_db --u=root --p=root
  `);

        log.info(`
  * createTable [tableName] [columns]

  column format:
  name:type:[index]:[nullable]:[default]

  * Details :
  name      Any
  type      Any postgresql type (or shortcuts : timestampz, int, bool)
  index     <unique|noindex> (optionnal)
  nullable  <null|notnull> (optionnal)
  default   <now()|true|false|...> (optionnal)

  Example :
  massive-collections createTable users \\
  username:varchar(255):unique:notnull \\
  password:varchar(255):noindex:notnull \\
  age:int \\
  details:jsonb \\
  created:timestampz:noindex:null:now() \\
  modified:timestampz:noindex:null:now()
  `)
      break;
    }
  }
}
catch(err) {
  log.error(err);
  log.error('Please read the documentation : massive-collections-cli help');
}

module.exports = {
  connect,
  disconnect,
  createTable,
  runQuery
};

// console.log(argv);
